File: js/Item.js
/**
 * The class for all the "items", the things that the player is supposed to catch.
 * (actually there will only be one item object created)
 */
class Item {
    constructor() {
        this.w = 15; // width
        this.h = 15; // height
        this.x = Way.w / 2;
        this.y = Way.h / 2;
        this.sprite = null; // Will be set later
    }
    
    // Getters
    getX() { return this.x; }
    getY() { return this.y; }
    getWidth() { return this.w; }
    getHeight() { return this.h; }
    
    // Setters
    setX(xx) { this.x = xx; }
    setY(yy) { this.y = yy; }
    setImage(itemSprite) { this.sprite = itemSprite; }
    
    /**
     * Draws the item
     */
    draw(ctx) {
        if (this.sprite) {
            ctx.drawImage(this.sprite, this.x, this.y);
        } else {
            // Fallback if image not loaded
            ctx.fillStyle = '#ff0000';
            ctx.fillRect(this.x, this.y, this.w, this.h);
        }
    }
    
    /**
     * Returns a rectangle bounds for collision detection
     */
    getBounds() {
        return {
            x: this.x,
            y: this.y,
            width: this.w,
            height: this.h
        };
    }
    
    /**
     * Moves the item to a new location
     */
    updatePosition(boxes) {
        let repeat = true;
        
        // Places the item at a random position
        // If it's placed on/behind a box, it's replaced
        while (repeat) {
            this.x = Math.floor(Math.random() * (Way.w - 11));
            this.y = Math.floor(Math.random() * (Way.h - 31)) + 20;
            
            repeat = false;
            for (let box of boxes) {
                if (box.checkHit(this.getBounds())) {
                    repeat = true;
                    break;
                }
            }
        }
    }
    
    /**
     * Checks collision with given bounds
     */
    checkHit(bounds) {
        const itemBounds = this.getBounds();
        return this.intersects(itemBounds, bounds);
    }
    
    /**
     * Helper method to check if two rectangles intersect
     */
    intersects(rect1, rect2) {
        return !(rect1.x + rect1.width < rect2.x ||
                rect2.x + rect2.width < rect1.x ||
                rect1.y + rect1.height < rect2.y ||
                rect2.y + rect2.height < rect1.y);
    }
}

File: Item.java
/**
The class for all the "items", the things that the player is supposed to catch.
(actually there will only be one item object created)
*/

import java.awt.*;

public class Item
{
	private int w, h; //The width and height of Item
	private int x, y; //The x and y coordinates Item
	private Image sprite; //The item sprite
	public Item()
	{
		w = 15;
		h = 15;

		x = Way.w/2;
		y = Way.h/2;
	}

	/**
	Get/set methods
	*/

	//Returns x
	public int getX(){return x;}

	//Returns y
	public int getY(){return y;}

	//Returns width
	public int getWidth(){return w;}

	//Returns height
	public int getHeight(){return h;}

	//Sets x
	public void setX(int xx){x = xx;}

	//Sets y
	public void setY(int yy){y = yy;}

	//Sets the sprite
	public void setImage(Image itemSprite)
	{
		sprite = itemSprite;
	}

	/**
	Draws the item
	*/
	public void draw(Graphics g)
	{
		g.drawImage(sprite, x, y, null);
	}

	//Returns a Rectangle the size of the item
	public Rectangle getBounds()
	{
		return new Rectangle(x, y, w, h);
	}

	/**
	Moves the item to a new location
	*/
	public void updatePosition(Box[] boxes)
	{
		boolean repeat = true;

		//Places the item at a random position
		//If it's placed on/behind a box, it's replaced
		while(repeat)
		{
			x = (int)(Math.random()*(Way.w-11));
			y = (int)(Math.random()*(Way.h-31)+20);

			repeat = false;
			for(Box tmp: boxes)
			{
				if(tmp.checkHit(getBounds())) repeat = true;
			}
		}
	}

	/**
	Checks collision
	*/
	public boolean checkHit( Rectangle bo )
	{
		boolean hit = getBounds().intersects( bo );

		return hit;
	}

}
